cf-custom-resources/lib/wkld-cert-validator.js

```diff
@@ -0,0 +1,881 @@
+        let unusedOptions = await unusedValidationOptions(
+          physicalResourceID,
+          loadBalancerDNS
+        );
+        await devalidate(unusedOptions);
+        await deleteCertificate(physicalResourceID);
+        break;
+      default:
+        throw new Error(`Unsupported request type ${event.RequestType}`);
+    }
+  };
+
+  try {
+    await Promise.race([exports.deadlineExpired(), handler()]);
+    await report(event, context, "SUCCESS", physicalResourceID);
+  } catch (err) {
+    console.log(`Caught error for service ${serviceName}: ${err.message}`);
+    await report(
+      event,
+      context,
+      "FAILED",
+      physicalResourceID,
+      null,
+      err.message
+    );
+  }
+};
+
+/**
+ * Delete the certificate.
+ *
+ * @param certARN The ARN of the certificate to delete.
+ * @returns {Promise<void>}
+ */
+async function deleteCertificate(certARN) {
+  // NOTE: wait for certificate to be not in-used.
+  let attempt;
+  for (attempt = 0; attempt < ATTEMPTS_CERTIFICATE_NOT_IN_USE; attempt++) {
+    let certificate;
+    try {
+      ({ Certificate: certificate } = await clients
+        .acm()
+        .describeCertificate({
+          CertificateArn: certARN,
+        })
+        .promise());
+    } catch (err) {
+      if (err.name === "ResourceNotFoundException") {
+        return;
+      }
+      throw err;
+    }
+
+    if (!certificate.InUseBy || certificate.InUseBy.length <= 0) {
+      break;
+    }
+    await sleep(30000);
+  }
+
+  if (attempt >= ATTEMPTS_CERTIFICATE_NOT_IN_USE) {
+    throw new Error(
+      `Certificate still in use after checking for ${ATTEMPTS_CERTIFICATE_NOT_IN_USE} attempts.`
+    );
+  }
+
+  await clients
+    .acm()
+    .deleteCertificate({ CertificateArn: certARN })
+    .promise()
+    .catch((err) => {
+      if (err.name !== "ResourceNotFoundException") {
+        throw err;
+      }
+    });
+}
+
+/**
+ * Validate that the aliases are not in use.
+ *
+ * @param {Set<String>} aliases for the service.
+ * @param {String} loadBalancerDNS the DNS of the service's load balancer.
+ * @throws error if at least one of the aliases is not valid.
+ */
+async function validateAliases(aliases, loadBalancerDNS) {
+  let promises = [];
+
+  for (let alias of aliases) {
+    let { hostedZoneID, route53Client } = await domainResources(alias);
+    const promise = route53Client
+      .listResourceRecordSets({
+        HostedZoneId: hostedZoneID,
+        MaxItems: "1",
+        StartRecordName: alias,
+        StartRecordType: "A",
+      })
+      .promise()
+      .then(({ ResourceRecordSets: recordSet }) => {
+        if (!targetRecordExists(alias, recordSet)) {
+          return;
+        }
+        if (recordSet[0].Type !== "A") {
+          return;
+        }
+        let aliasTarget = recordSet[0].AliasTarget;
+        if (
+          aliasTarget &&
+          loadBalancerDNS &&
+          aliasTarget.DNSName.toLowerCase() ===
+            `${loadBalancerDNS.toLowerCase()}.`
```

Comment:
nit: not a big deal, but I feel like sometimes it is easier to read when things are in the same line 🤔 
```suggestion
          aliasTarget.DNSName.toLowerCase() === `${loadBalancerDNS.toLowerCase()}.`
```